### ISSUE TITLE
PDF Preview not working

### ISSUE THREAD

**Issue by ampaze** [sulu/sulu#6376]
| Q | A
| --- | ---
| Sulu Version | 2.2.17
| PHP Version | 8.0

# Actual Behavior

PDFs don't have a preview image in PHP 8. 
My guess is: as there is no Imagick yet for PHP 8, Imagine has to use Gd, which does not support pdf.

# Expected Behavior

That a preview image of the pdf is generated.

# Steps to Reproduce

Upload a PDF to Sulu and see that no preview image is generated.

# Possible Solutions

As far as I was able to analyse this, the problem is in FormatManager.php, `getSupportedOutputImageFormats` accepts `application/pdf` but does not return `pdf` as a supported image format. So convert is never called. 

```php
$supportedImageFormats = $this->converter->getSupportedOutputImageFormats($fileVersion->getMimeType());
if (empty($supportedImageFormats)) {
    throw new InvalidMimeTypeForPreviewException($fileVersion->getMimeType());
}

if (!\in_array($imageFormat, $supportedImageFormats)) {
    throw new ImageProxyInvalidImageFormat(/*...*/);
}

// Convert Media to format.
$responseContent = $this->converter->convert($fileVersion, $formatKey, $imageFormat);
```

$fileVersion->getMimeType() == `application/pdf`
$imageFormat == `pdf`
$supportedImageFormats == `array(4) { [0]=> string(3) "jpg" [2]=> string(3) "gif" [3]=> string(3) "png" [4]=> string(4) "webp" }`

Furthermore, if you remove the imageFormat check, it seems to work, but the next error is

> Saving image in "pdf" format is not supported, please use one of the following extensions: "bmp", "gif", "jpeg", "png", "wbmp", "webp", "xbm"

----

The most simple solution would be, change

```php
$imageFormat = $info['extension'];
```

to

```php
$imageFormat = $info['extension'] == 'pdf' ? 'jpg' : $info['extension'];
```

This needs of course to have ghost script configured
```yml
sulu_media:
    ghost_script:
        path: xxxxx
```

**Comment by alexander-schranz** [comment]
@ampaze The `$imageFormat` is false, looks like you accidently type into your url `.pdf` instead of `.jpg`. Where did you get that url from, how did you build the url of your thumbnails? 

The FormatManager works for me like expected if you have configured `ghost_script` the `getSupportedOutputImageFormats` will return the image formtats supported for pdfs and so you will be possible to generate `jpg`, `png`, `webp`, `gif` previews for PDF Documents. Don't get confused by `supportedImageFormats` not containing pdf as `pdf` is not the target format sulu supports. The source mimetype can be pdf but the targets will only be image formats like listed above.

As an example when uploading the following pdf to `sulu.rocks` you will see the following preview:

[sample.pdf](https://github.com/sulu/sulu/files/7587448/sample.pdf)

<img width="317" alt="Bildschirmfoto 2021-11-23 um 11 59 49" src="https://user-images.githubusercontent.com/1698337/143012823-2a7c1bc3-8904-465e-8c80-f004b0ebbe59.png">

If previews / image formats don't work its most common that ghostscript is not configured correctly or the user running php has no access to ghostscript.

But when your `$imageFormat` variable is `pdf` something wrong with the url. As image formats url are always `jpg`, `webp`, ... or the other supported image extensions.

**Comment by ampaze** [comment]
Thanks for the detailed reply. 👍 

In debugging this issue I accidentally left the `pdf` file extension in the url.

The real problem was that `image.thumbnails` is empty for pdfs when `ghost_script` is not configured. 

I configured it only after I had already changed stuff around. And that made it work, not my changes. Sorry for that.

----

Still, I only found out about `ghost_script` in others issues here on github. It is not documented at all, or it least not where could find it. I imagine this will be a problem for all users using PHP8.
